### ISSUE TITLE
[Documentation]: VS Code integration with the direnv extension

### ISSUE THREAD

**Issue by apgrucza** [jetpack-io/devbox#1029]
The Devbox documentation makes mention of the Devbox VS Code extension and dev containers, but neither currently provide VS Code users with an adequate experience (as described in #1001). In that issue it was identified that the [direnv VS Code extension](https://marketplace.visualstudio.com/items?itemName=mkhl.direnv) provides a more streamlined Devbox experience. I find this approach preferable to dev containers, which negate the benefits of Devbox as a Docker-free solution.

I wish I'd come across the direnv extension earlier, as I spent a lot of time looking for workarounds and nearly gave up on Devbox. Having it in the Devbox documentation could prevent others having the same experience.

Note the [known issues](https://marketplace.visualstudio.com/items?itemName=mkhl.direnv#known-issues) with this extension. There are two other direnv extensions in the marketplace, but I've not tried them.

**Comment by Lagoja** [comment]
@apgrucza we had some struggles with the direnv extension + Devbox in the past, but if it's working well we can consider recommending it. Does your project work well since you installed it?

We've made changes to our direnv integration, so it's possible that we're now at a state where they work well together

**Comment by Lagoja** [comment]
Did some testing with a few example projects, and this seems to work reasonably well if a .envrc file is present. There are some limitations highlighted in the extension's docs that are worth noting. 

- The Direnv Extension doesn't seem to work with tasks of `process` type
- There seem to be reports where the Direnv Extension does not set the environment quickly enough for other extensions to detect and load the new environment

I still think it's worth recommending as an option. Some next steps we could take on this issue: 

1. Add a note about the Direnv extension to the Direnv + VSCode pages in our documentation
2. Test the IntelliJ Direnv extension to see if that also helps with IntelliJ setup

I would also still like to find a way for Devbox to work with VSCode and other editors without having dependencies on direnv + the direnv extension.

**Comment by apgrucza** [comment]
The direnv extension worked well for my Devbox trial, but I didn't do extensive testing (I tried the terminal, a task of type shell, Python debugging and an extension that tries to execute a locally installed tool).

To remove the dependency on direnv, perhaps some functionality could be copied from the direnv extension into the Devbox extension.

Good IDE integration is pretty important in ensuring Devbox gains wider adoption. If Devbox can work with both VS Code and IntelliJ IDEs, that might make it a more attractive option than Dev Containers (which are currently VSCode-only). But [this IntelliJ issue](https://youtrack.jetbrains.com/issue/IDEA-320397) suggests that the direnv plugins don't work particularly well.

**Comment by mohsenari** [comment]
@apgrucza We're looking into how direnv extension is handling environment management to see if we can implement setting environment in devbox extension as well. 

The main issue is we can't be sure ***when*** vscode language extensions (golang, python, java etc.) read the env variables such as `$GOPATH`, `$PYTHONPATH`, or `$JAVAHOME` etc. because devbox extension needs to set or override those env variables ***before*** they're read by a language extension. The other issue is, we don't know how it affects debugging and vscode tasks.

So the ideal case is if the user runs `devbox shell` and then `code .` to open vscode from inside the devbox environment. Or alternatively, if direnv is setup, `cd project_dir` and then `code .` because that way, the vscode process starts in a devbox environment before any of its extensions are loaded. 
But that doesn't usually happen to be the case. Usually, you have vscode open and you open the workspace directory from within vscode. In that case, the best an extension can do is to play with `process.env` and that's what direnv extension does (as far as I know). So right now we're testing updating `process.env` to see how it affects vscodes integration with a devbox environment.

If possible, would you mind sharing the tests you ran for tasks and python debugging?

**Comment by mohsenari** [comment]
For more context around this very issue, see this [thread](https://github.com/microsoft/vscode/issues/152806).

**Comment by apgrucza** [comment]
@mohsenari See #1028 for the test I did with tasks.

For Python debugging, I used the [Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python).

In `devbox.json`:

```json
{
  "packages": [
    "python@3.10",
    "python310Packages.pip@23"
  ],
  "shell": {
    "init_hook": "source $VENV_DIR/bin/activate"
  },
  "nixpkgs": {
    "commit": "f80ac848e3d6f0c12c52758c0f25c10c97ca3b62"
  }
}
```

In `.vscode/settings.json`:

```json
{
    "python.defaultInterpreterPath": "${env:VENV_DIR}/bin/python"
}
```

In `.vscode/launch.json`:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "justMyCode": true
        }
   ]
}
```

Then I opened a .py file, added a breakpoint with F9, and started debugging with F5.

**Comment by mohsenari** [comment]
@apgrucza I'm just putting a comment here to give a small update. 
I've been working on adding a better integration between devbox and vscode that works better than direnv extension. 
The PR is out but it's still needs a bit of cleanup and review. But to summarize, it allows for changing the process environment in which vscode runs to a devbox environment as if you opened vscode from a devbox shell. I'll ping here again once the PR is merged and the feature is out.

**Comment by apgrucza** [comment]
> @apgrucza I'm just putting a comment here to give a small update. I've been working on adding a better integration between devbox and vscode that works better than direnv extension. The PR is out but it's still needs a bit of cleanup and review. But to summarize, it allows for changing the process environment in which vscode runs to a devbox environment as if you opened vscode from a devbox shell. I'll ping here again once the PR is merged and the feature is out.

Thanks for the update, it sounds promising. Are you planning to have a setting in the extension for enabling the new "pure" mode? Or perhaps there should be a setting in `devbox.json` to enable pure mode for all users, as suggested by @codethief [over here](https://github.com/jetpack-io/devbox/issues/403#issuecomment-1365149061).

**Comment by mohsenari** [comment]
@apgrucza oh both are very good suggestions, I haven't considered that. I have to do some testing to see how it works in vscode. Let me discuss with the team to see where we land on it.

**Comment by mohsenari** [comment]
@apgrucza we just published a new version of devbox VSCode extension that has the ability to reopen VSCode in a devbox environment.
Unlike direnv extension, we change the environment variables of the vscode `process` (hence the reopening). That means other extensions and vscode settings will point to your devbox packages if they look for an executable (e.g., python). It was a challenging task to overcome, but it is pretty much the same as opening vscode from a terminal that has devbox shell enabled.

Here is [the link to the extension](https://marketplace.visualstudio.com/items?itemName=jetpack-io.devbox), but you can install it via VSCode extension search when you search for "jetpack devbox".
Please give it a try by doing these steps and let us know if you have any feedback:
1. opening a project in vscode that has a devbox.json
2. opening command palette (cmd+shift+p) and searching for "reopen in devbox"
3. once vscode is reopen, confirm if env variables in terminal are set correctly to your devbox environment. (it updates the env variables not just in terminal but also in all of vscode, but it's harder to check env variables outside of terminal in vscode).

**Comment by apgrucza** [comment]
@mohsenari not sure if you want the feedback here or in a separate issue.

The first time I selected "reopen in devbox", I saw the popup saying that it was installing packages, and then the VS Code window closed before it had finished and did not reopen. Now when I try again, the VS Code window closes immediately and also does not reopen. It seems like it might be crashing.

I am using WSL - I wonder if that might be related.

**Comment by apgrucza** [comment]
> Thanks for the update, it sounds promising. Are you planning to have a setting in the extension for enabling the new "pure" mode? Or perhaps there should be a setting in `devbox.json` to enable pure mode for all users, as suggested by @codethief [over here](https://github.com/jetpack-io/devbox/issues/403#issuecomment-1365149061).

@mohsenari I've created #1226 to address this, but feel free to suggest alternatives if it doesn't align with what you've discussed with the team.

**Comment by apgrucza** [comment]
@mohsenari how did you go with the WSL support? Was the problem with launching VS Code, or with passing environment variables through? I think both should be achievable. For example, try running the following in a WSL window:

```bash
FOO=bar code
```

VS Code will open, and if you open a terminal and type `echo $FOO` you will see the value of `bar` has carried through.

**Comment by mohsenari** [comment]
hey @apgrucza thanks for the follow up. I noticed that too. Even from windows CMD you can run a command in wsl by doing `wsl <command>` so I added that feature to the extension so that when you run `Reopen in Devbox shell` it would run a devbox command in wsl in terminal. And that command would setup the env variables and would open vscode from inside wsl. 
But there were 2 main problems:
1. I couldn't establish a process communication channel (ipc) between vscode in windows and the devbox command that's running in WSL. That meant devbox couldn't signal to vscode for closing the window. So you'd end up with 2 open vscode windows.
2. The new opened VSCode window had the correct environment except for PATH. So for example, if you have `hello` package installed in your devbox.json, in the new terminal, if you run `which hello` it doesn't point to your devbox package. 
That means people will need to run `devbox shell` in their terminal even if they use the `Reopen in Devbox shell` command.

So because of these 2 main hurdles, I paused the development on this for now to focus on other tasks, and put an error message with link to documentation for manual setup. I can pick it back up in the next couple weeks if there is enough interest.

**Comment by apgrucza** [comment]
Thanks for the detailed response @mohsenari. I had a poke around myself in `internal/boxcli/integrate.go` and made these observations:

- The IPC is working ok (it's able to get the config dir and terminate the parent VS Code process).
- The `code` command actually runs a shell script from VS Code Server rather than the Windows VS Code executable. This is because VS Code has modified `PATH` to include an entry like `/home/XXXX/.vscode-server/bin/XXXX/bin/remote-cli`.
- The `code` shell script runs JavaScript code which fails to start a new VS Code process because the parent VS Code process has been terminated.
- If the parent VS Code process is kept alive, a new VS Code process is successfully created but the JavaScript code does not forward environment variables to it.

If `HOME` is set, the Windows executable can be run instead, and environment variables do get passed through. So I'm not sure if this is the best way, but the below changes to `integrate.go` do work as a proof-of-concept under WSL:

```go
	cmnd := exec.Command(os.Getenv("VSCODE_CWD")+"/bin/code", message.ConfigDir)
	cmnd.Env = append(cmnd.Env, envVars...)
	cmnd.Env = append(cmnd.Env, "HOME="+os.Getenv("HOME")) // Prevent error: mkdir: cannot create directory ‘/.vscode-server’: Permission denied
	cmnd.Env = append(cmnd.Env, "ELECTRON_RUN_AS_NODE=")   // Prevent error: Remote Extension host terminated unexpectedly 3 times within the last 5 minutes.
```

Would it help if a created a separate issue for this (Add WSL support to the Devbox VS Code extension). This issue can then be closed, as the documentation change appears to be complete.

**Comment by mohsenari** [comment]
@apgrucza thanks for looking into this. As you suggested please create a separate issue for the WSL use case. We can discuss the details there.

**Comment by apgrucza** [comment]
Fixed by #1080

**Comment by codethief** [comment]
@apgrucza 

> Are you planning to have a setting in the extension for enabling the new "pure" mode? Or perhaps there should be a setting in `devbox.json` to enable pure mode for all users, as suggested by @codethief [over here](https://github.com/jetpack-io/devbox/issues/403#issuecomment-1365149061).

I'm a bit late to the party (thanks for linking me!) but are there any docs about this new "pure" mode? This is the first time I'm hearing about it. The only thing I've been able to find is the [release note](https://github.com/jetpack-io/devbox/releases/tag/0.5.5).

**Comment by apgrucza** [comment]
@codethief, I think it's undocumented at the moment because it's still an experimental feature - see https://github.com/jetpack-io/devbox/issues/1226#issuecomment-1611885514.

**Comment by mohsenari** [comment]
@codethief @apgrucza my apologies on this, I must've forgotten to update the docs when we added this feature. FWIW, if you run `devbox shell --help` it will show the flag as an option there. Just not the devbox docs webpage for which I created a PR to update.
